text/projects/Lukasz_Selwa/Galil_Seiferas.tex

```diff
@@ -0,0 +1,308 @@
+
+
+W pracy ,,Time-Space-Optimal String Matching'' Z. Galil i J. Seiferas jako pierwsi przedstawili algorytm do wyszukiwania wzorca działający w czasie liniowym i stałej pamięci.
+Niniejszy tekst jest oparty na późniejszej pracy M. Crochemore i W. Rytter ,,Squares, Cubes, and Time-Space Efficient StringS earching'', która zawiera uproszczony opis i dowód oryginalnego algorytmu Galil-Seiferas.
+
+\section{Algorytm}
+
+\subsection{Idea algorytmu}
+Idea algorytmu Galil-Seiferas polega na zmodyfikowaniu działania algorytmu Morris–Pratta tak aby oszczędzić trzymania w pamięci całej tablicy prefikso-sufiksów.
+Dokonuje się tego poprzez dekompozycje wzorca $w$ na słowa $u,v$, takie że $w = u v$, $u$ jest stosunkowo krótkim słowem a $v$ nie zawiera dużych prefiksów.
+W słowie $t$ szukając wystąpień wzorca $w$ będziemy znajdować wystąpienia $v$ i dla każdego wystąpienia naiwnie sprawdzać czy bezpośrednio przed $v$ występuje $u$.
+Okazuję się, że da się to zrobić tak, żeby cały algorytm dalej działał w czasie liniowym od $|t| + |w|$.
+
+\subsection{Opis Algorytmu}
+
+\begin{definition}
+\textbf{\textit{k-HRP}}.
+Dla $k \geq 3$ mówimy, że słowo $v$ jest $k$-wysoce-powtarzanym-prefiksem ($k$-HRP, ang. $k$-\textit{highly-repeating-prefix}) słowa $w$, jeśli $v^i$ jest prefiksem $w$ dla pewnego $i \geq k$ oraz $v$ jest słowem pierwotnym.
+Będziemy oznaczać $k$-HRP przez samo HRP jeśli $k$ będzie jasne z kontekstu.
+\end{definition}
+
+Zauważmy, że najkrótszy okres słowa $v^2$ to $|v|$. Inaczej $v$ nie byłoby słowem pierwotnym. Z definicji $k$-HRP $v^2$ jest prefiksem $w$. Dzięki temu możemy zdefiniować zakres $v$.
+
+\begin{definition}
+\textbf{\textit{Zakres.}}
+Niech $v$ będzie $k$-HRP słowa $w$. Przedział $[L, R]$ nazywamy zakresem $v$ jeśli $L = |v^2|$ oraz $[L,R]$ jest najdłuższym przedziałem takim, że dla każdego $i \in [L, R]$ najkrótszy okres $w[1...i]$ jest równy $|v|$.
+\end{definition}
+
+\begin{lemma}\label{Lem_1}
+Niech $v_1, v_2$ to $k$-HRP słowa $w$, $|v_1| < |v_2|$ oraz $[L_1, R_1], [L_2, R_2]$ to zakresy odpowiednio $v_1, v_2$. Wtedy $R_1 < L_2$.
+\end{lemma}
+\begin{proof}
+Niech $z$ będzie prefiksem $w$ długości $R_1$. Jeśli $L_2 \leq R_1$ to z tego wynika, że $v_2^2$ jest prefiksem $z$. Ponieważ, $|v_1| < |v_2|$ i $|v_1|$ jest okresem $z = v_2^2 z'$ to $|v_1|$ jest też okresem $v_2^2$. Z lematu o okresowości wynika, że $\gcd(|v_1|, |v_2|)$ też musi być okresem $v_2^2$. Ale $\gcd(|v_1|, |v_2|) < |v_2|$ stąd sprzeczność z pierwotnością $v_2$.
+\end{proof}
+
+Dla słowa $x$, przez $p(x)$ oznaczamy najkrótszy okres $x$.
+
+\begin{lemma}\label{Lem_2}
+Niech $v_1, ..., v_r$ będzie sekwencją wszystkich $k$-HRP słowa $w$. Niech $[L_i, R_i]$ to zakres $v_i$. Wtedy dla każdego niepustego prefiksu $u$ słowa $w$ zachodzi:
+$$\begin{array}{rl}
+    p(u) = L_i/2 & \text{jeśli $|u| \in [L_i, R_i]$ dla pewnego $i$} \\
+    p(u) > |u|/k & \text{w przeciwnym przypadku}
```

Comment:
p(u) = 
\begin{cases}
L_i/2 & \text{jeśli $|u| \in [L_i, R_i]$ dla pewnego $i$} \\
|u|/k & \text{w przeciwnym przypadku}
\end{cases}